client/normal_value.go

```diff
@@ -0,0 +1,591 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package client
+
+import (
+	"time"
+
+	"github.com/sourcenetwork/immutable"
+	"golang.org/x/exp/constraints"
+)
+
+type NormalValue struct {
+	value any
```

Comment:
suggestion: It might be worth chatting over discord about this before commiting too much time to polishing it, I have a few concerns that need talking over and others may have some too.  My biggest are below:

1. This means a pointer for every field value
2. Loss of compile time safety

You might actually be able to remove the second concern by using a constrained generic instead of `any`, in most langs this would also resolve (1), but IIRC in Go all generic variables are pointers anyway (I'm not sure, worth checking):
```
type NormalType interface {
  int64 | float64 | string | etc
}

type NormalValue[T NormalType]  struct {
  value T
}
```

Reply:
> in Go all generic variables are pointers anyway

I don't think that's true. Do you remember where you might have seen this?

Reply:
> It might be worth chatting over discord

I actually wanted to bring it up yesterday, but the time switch caught me off the guard. And then we want over the time.

> Loss of compile time safety

We use `any` everywhere and so far it wasn't a concern.
In this case `NormalValue` is meant to provide some type safety in more compact and reliable way.
It it mostly a wrapper over the value held by `client.FieldValue` which is also of `any` type.

Reply:
> It is mostly a wrapper over the value held by client.FieldValue which is also of any type.

Could we not apply all the `NormalValue` methods directly to `FieldValue`?

Reply:
> in Go all generic variables are pointers anyway

I might need to do some additional research on this, but I hardly believe.
In any language with decent support of generics (like C++ or Rust) they are all resolved at compile time and every instantiation of a generic type is a distinct type.

Reply:
https://deepsource.com/blog/go-1-18-generics-implementation

> In Boxing, "values" are boxed and passed around as references of the polymorphic type, generally using a pointer table, commonly called virtual table. This is how Go's interface are implemented. This generally produces a smaller binary and requires short compile-times but may have a toll on the runtime performance.
> 
> Go generics combines concepts from "monomorphisation" (stenciling) and "boxing" (dynamic dispatch) and is implemented using [GCshape stenciling and dictionaries](https://github.com/golang/proposal/blob/master/design/generics-implementation-dictionaries-go1.18.md). This allows Go to have fast compile times and smaller binaries while having generics.

Go generics are not the same generics that you find in Rust/C#/C++/etc - they have a runtime cost here :(

Reply:
> We use any everywhere and so far it wasn't a concern.

This is true, when I think about it our code is all very reliant on boxing anyway, however I think the proposed code might result in double boxing - e.g. in a `core.Doc.Fields` we have an array of `any`, which here would box `NormalValue`, which in turn boxes the field value.

If we are able to change the type of `core.Doc.Fields` to `NormalValue` we might only have a single box (my preference being the generic, not an any anyway).